src/common/datasource/src/file_format/orc.rs

```diff
@@ -91,16 +102,21 @@ impl<T: AsyncRead + AsyncSeek + Unpin + Send + 'static> Stream for OrcArrowStrea
 
         let batch = batch.map(|b| {
             b.and_then(|b| {
-                let mut columns = Vec::with_capacity(b.num_columns());
-                for (idx, column) in b.columns().iter().enumerate() {
-                    if column.data_type() != self.output_schema.field(idx).data_type() {
-                        let output = cast(&column, self.output_schema.field(idx).data_type())?;
+                let mut columns = Vec::with_capacity(self.projection.len());
+                for idx in self.projection.iter() {
+                    let column = b.column(*idx);
+                    let field = self.output_schema.field(*idx);
+
+                    if column.data_type() != field.data_type() {
+                        let output = cast(&column, field.data_type())?;
                         columns.push(output)
                     } else {
                         columns.push(column.clone())
                     }
                 }
-                let record_batch = DfRecordBatch::try_new(self.output_schema.clone(), columns)?;
+
+                let projected_schema = self.output_schema.project(&self.projection)?;
```

Comment:
Move L118 before the L105. Therefore, we can safely do operations like `b.column(*idx)` without out-of-bound error.